### PR TITLE
Add missing externs for Array

### PR DESCRIPTION
**Pull request by simon0191**
- Bump version to 0.3.0
- Add unshift, shift, dup, to_s and reverse to Array

### CODE REVIEW

File: src/array.rs [hunk 1]
```diff
@@ -11,6 +11,11 @@ extern "C" {
     pub fn rb_ary_push(array: Value, item: Value) -> Value;
     pub fn rb_ary_store(array: Value, index: c_long, item: Value) -> Value;
     pub fn rb_ary_pop(array: Value) -> Value;
+    pub fn rb_ary_unshift(array: Value, item: Value) -> Value;
+    pub fn rb_ary_shift(array: Value) -> Value;
+    pub fn rb_ary_dup(array: Value) -> Value;
+    pub fn rb_ary_to_s(array: Value) -> Value;
+    pub fn rb_ary_reverse(array: Value) -> Value;
```
Comment (malept) [inline src/array.rs]: Perhaps all of these declarations should be alphabetized.

**Comment by malept** [comment]
Perhaps while we're doing this we should also add `rb_ary_sort`, `rb_ary_sort_bang`, `rb_ary_concat`, etc.

**Comment by simon0191** [comment]
Done @malept

File: Cargo.toml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "ruby-sys"
-version = "0.2.13"
+version = "0.3.0"
```
Comment (d-unsed) [inline Cargo.toml]: Please don't include the version change to the PR. I'll bump it myself before releasing.

File: src/array.rs [hunk 3]
```diff
@@ -5,12 +5,20 @@ use constant::{FL_USHIFT, FL_USER_1, FL_USER_3, FL_USER_4};
 use types::{c_long, InternalValue, RBasic, Value};
 
 extern "C" {
+    pub fn rb_ary_concat(array: Value,other_array: Value) -> Value;
```
Comment (d-unsed) [inline src/array.rs]: Missing whitespace after `,`

File: Cargo.toml (hunk 2 above)
Comment (malept) [reply Cargo.toml]: @d-unseductable I suggest adding this to (a currently non-existent) `CONTRIBUTING.md` (I need to add something about version bumping in all of my projects too...)

**Comment by nomoon** [comment]
Been following this. Not sure if the following is helpful, but here are basically _all_ of the `rb_*` functions in Ruby's array.c, ruby-sys/rust-ified (though not really "proof-read" for correctness or usefulness) with their original variable names. In case there are any more that stick out as pressing for inclusion:

```
    pub fn rb_mem_clear(mem: *Value, size:  c_long);
    pub fn rb_ary_decrement_share(shared: Value);
    pub fn rb_ary_unshare(ary: Value);
    pub fn rb_ary_unshare_safe(ary: Value);
    pub fn rb_ary_increment_share(shared: Value) -> Value;
    pub fn rb_ary_set_shared(ary: Value, shared: Value);
    pub fn rb_ary_modify_check(ary: Value);
    pub fn rb_ary_modify(ary: Value);
    pub fn rb_ary_freeze(ary: Value) -> Value;
    pub fn rb_ary_frozen_p(ary: Value) -> Value;
    pub fn rb_ary_shared_with_p(ary1: Value, ary2: Value) -> Value;
    pub fn rb_ary_new_capa(capa: c_long) -> Value;
    pub fn rb_ary_new() -> Value;
    pub fn rb_ary_new_from_args(n: c_long, ...) -> Value;
    pub fn rb_ary_new_from_values(n: c_long, elts: *const Value) -> Value;
    pub fn rb_ary_tmp_new(capa: c_long) -> Value;
    pub fn rb_ary_tmp_new_fill(capa: c_long) -> Value;
    pub fn rb_ary_free(ary: Value);
    pub fn rb_assoc_new(car: Value, cdr: Value) -> Value;
    pub fn rb_check_array_type(ary: Value) -> Value;
    pub fn rb_ary_s_try_convert(dummy: Value, ary: Value) -> Value;
    pub fn rb_ary_initialize(argc: c_int, argv: *Value, ary: Value) -> Value;
    pub fn rb_ary_s_create(argc: c_int, argv: *Value, klass: Value) -> Value;
    pub fn rb_ary_store(ary: Value, idx: c_long, val: Value);
    pub fn rb_ary_push(ary: Value, item: Value) -> Value;
    pub fn rb_ary_cat(ary: Value, argv: *const Value, len: c_long) -> Value;
    pub fn rb_ary_push_m(argc: c_int, argv: *Value, ary: Value) -> Value;
    pub fn rb_ary_pop(ary: Value) -> Value;
    pub fn rb_ary_pop_m(argc: c_int, argv: *Value, ary: Value) -> Value;
    pub fn rb_ary_shift(ary: Value) -> Value;
    pub fn rb_ary_shift_m(argc: c_int, argv: *Value, ary: Value) -> Value;
    pub fn rb_ary_unshift_m(argc: c_int, argv: *Value, ary: Value) -> Value;
    pub fn rb_ary_unshift(ary: Value, item: Value) -> Value;
    pub fn rb_ary_elt(ary: Value, offset: c_long) -> Value;
    pub fn rb_ary_entry(ary: Value, offset: c_long) -> Value;
    pub fn rb_ary_subseq(ary: Value, beg: c_long, len: c_long) -> Value;
    pub fn rb_ary_aref(argc: c_int, argv: *const Value, ary: Value) -> Value;
    pub fn rb_ary_at(ary: Value, pos: Value) -> Value;
    pub fn rb_ary_first(argc: c_int, argv: *Value, ary: Value) -> Value;
    pub fn rb_ary_last(argc: c_int, argv: *const Value, ary: Value) -> Value;
    pub fn rb_ary_fetch(argc: c_int, argv: *Value, ary: Value) -> Value;
    pub fn rb_ary_index(argc: c_int, argv: *Value, ary: Value) -> Value;
    pub fn rb_ary_rindex(argc: c_int, argv: *Value, ary: Value) -> Value;
    pub fn rb_ary_to_ary(obj: Value) -> Value;
    pub fn rb_ary_splice(ary: Value, beg: c_long, len: c_long, rptr: *const Value, rlen: c_long);
    pub fn rb_ary_set_len(ary: Value, len: c_long);
    pub fn rb_ary_resize(ary: Value, len: c_long) -> Value;
    pub fn rb_ary_aset(argc: c_int, argv: *Value, ary: Value) -> Value;
    pub fn rb_ary_insert(argc: c_int, argv: *Value, ary: Value) -> Value;
    pub fn rb_ary_length(ary: Value); -> Value;
    pub fn rb_ary_each(ary: Value) -> Value;
    pub fn rb_ary_each_index(ary: Value) -> Value;
    pub fn rb_ary_reverse_each(ary: Value) -> Value;
    pub fn rb_ary_length(ary: Value) -> Value;
    pub fn rb_ary_empty_p(ary: Value) -> Value;
    pub fn rb_ary_dup(ary: Value) -> Value;
    pub fn rb_ary_resurrect(ary: Value) -> Value;
    pub fn rb_ary_join(ary: Value, sep: Value) -> Value;
    pub fn rb_ary_join_m(argc: c_int, argv: *Value, ary: Value) -> Value;
    pub fn rb_ary_inspect(ary: Value) -> Value;
    pub fn rb_ary_to_s(ary: Value) -> Value;
    pub fn rb_ary_to_a(ary: Value) -> Value;
    pub fn rb_ary_to_h(ary: Value) -> Value;
    pub fn rb_ary_to_ary_m(ary: Value) -> Value;
    pub fn rb_ary_reverse(ary: Value) -> Value;
    pub fn rb_ary_reverse_bang(ary: Value) -> Value;
    pub fn rb_ary_reverse_m(ary: Value) -> Value;
    pub fn rb_ary_rotate(ary: Value, cnt: c_long) -> Value;
    pub fn rb_ary_rotate_bang(argc: c_int, argv: *Value, ary: Value) -> Value;
    pub fn rb_ary_rotate_m(argc: c_int, argv: *Value, ary: Value) -> Value;
    pub fn rb_ary_sort_bang(ary: Value) -> Value;
    pub fn rb_ary_sort(ary: Value) -> Value;
    pub fn rb_ary_bsearch_index(ary: Value); -> Value;
    pub fn rb_ary_bsearch(ary: Value) -> Value;
    pub fn rb_ary_bsearch_index(ary: Value) -> Value;
    pub fn rb_ary_sort_by_bang(ary: Value) -> Value;
    pub fn rb_ary_collect(ary: Value) -> Value;
    pub fn rb_ary_collect_bang(ary: Value) -> Value;
    pub fn rb_get_values_at(obj: Value, olen: c_long, argc: c_int, argv: *const Value, func: CallbackPtr) -> Value;
    pub fn rb_ary_values_at(argc: c_int, argv: *Value, ary: Value) -> Value;
    pub fn rb_ary_select(ary: Value) -> Value;
    pub fn rb_ary_select_bang(ary: Value) -> Value;
    pub fn rb_ary_keep_if(ary: Value) -> Value;
    pub fn rb_ary_delete(ary: Value, item: Value) -> Value;
    pub fn rb_ary_delete_same(ary: Value, item: Value);
    pub fn rb_ary_delete_at(ary: Value, pos: c_long) -> Value;
    pub fn rb_ary_delete_at_m(ary: Value, pos: Value) -> Value;
    pub fn rb_ary_slice_bang(argc: c_int, argv: *Value, ary: Value) -> Value;
    pub fn rb_ary_reject_bang(ary: Value) -> Value;
    pub fn rb_ary_reject(ary: Value) -> Value;
    pub fn rb_ary_delete_if(ary: Value) -> Value;
    pub fn rb_ary_zip(argc: c_int, argv: *Value, ary: Value) -> Value;
    pub fn rb_ary_transpose(ary: Value) -> Value;
    pub fn rb_ary_replace(copy: Value, orig: Value) -> Value;
    pub fn rb_ary_clear(ary: Value) -> Value;
    pub fn rb_ary_fill(argc: c_int, argv: *Value, ary: Value) -> Value;
    pub fn rb_ary_plus(x: Value, y: Value) -> Value;
    pub fn rb_ary_concat_multi(argc: c_int, argv: *Value, ary: Value) -> Value;
    pub fn rb_ary_concat(x: Value, y: Value) -> Value;
    pub fn rb_ary_times(ary: Value, times: Value) -> Value;
    pub fn rb_ary_assoc(ary: Value, key: Value) -> Value;
    pub fn rb_ary_rassoc(ary: Value, value: Value) -> Value;
    pub fn rb_ary_equal(ary1: Value, ary2: Value) -> Value;
    pub fn rb_ary_eql(ary1: Value, ary2: Value) -> Value;
    pub fn rb_ary_hash(ary: Value) -> Value;
    pub fn rb_ary_includes(ary: Value, item: Value) -> Value;
    pub fn rb_ary_cmp(ary1: Value, ary2: Value) -> Value;
    pub fn rb_ary_diff(ary1: Value, ary2: Value) -> Value;
    pub fn rb_ary_and(ary1: Value, ary2: Value) -> Value;
    pub fn rb_ary_or(ary1: Value, ary2: Value) -> Value;
    pub fn rb_ary_max(argc: c_int, argv: *Value, ary: Value) -> Value;
    pub fn rb_ary_min(argc: c_int, argv: *Value, ary: Value) -> Value;
    pub fn rb_ary_uniq_bang(ary: Value) -> Value;
    pub fn rb_ary_uniq(ary: Value) -> Value;
    pub fn rb_ary_compact_bang(ary: Value) -> Value;
    pub fn rb_ary_compact(ary: Value) -> Value;
    pub fn rb_ary_count(argc: c_int, argv: *Value, ary: Value) -> Value;
    pub fn rb_ary_flatten_bang(argc: c_int, argv: *Value, ary: Value) -> Value;
    pub fn rb_ary_flatten(argc: c_int, argv: *Value, ary: Value) -> Value;
    pub fn rb_ary_shuffle_bang(argc: c_int, argv: *Value, ary: Value) -> Value;
    pub fn rb_ary_shuffle(argc: c_int, argv: *Value, ary: Value) -> Value;
    pub fn rb_ary_sample(argc: c_int, argv: *Value, ary: Value) -> Value;
    pub fn rb_ary_cycle_size(self: Value, args: Value, eobj: Value) -> Value;
    pub fn rb_ary_cycle(argc: c_int, argv: *Value, ary: Value) -> Value;
    pub fn rb_ary_permutation_size(ary: Value, args: Value, eobj: Value) -> Value;
    pub fn rb_ary_permutation(argc: c_int, argv: *Value, ary: Value) -> Value;
    pub fn rb_ary_combination_size(ary: Value, args: Value, eobj: Value) -> Value;
    pub fn rb_ary_combination(ary: Value, num: Value) -> Value;
    pub fn rb_ary_repeated_permutation_size(ary: Value, args: Value, eobj: Value) -> Value;
    pub fn rb_ary_repeated_permutation(ary: Value, num: Value) -> Value;
    pub fn rb_ary_repeated_combination_size(ary: Value, args: Value, eobj: Value) -> Value;
    pub fn rb_ary_repeated_combination(ary: Value, num: Value) -> Value;
    pub fn rb_ary_product(argc: c_int, argv: *Value, ary: Value) -> Value;
    pub fn rb_ary_take(obj: Value, n: Value) -> Value;
    pub fn rb_ary_take_while(ary: Value) -> Value;
    pub fn rb_ary_drop(ary: Value, n: Value) -> Value;
    pub fn rb_ary_drop_while(ary: Value) -> Value;
    pub fn rb_ary_any_p(ary: Value) -> Value;
    pub fn rb_ary_dig(argc: c_int, argv: *Value, self: Value) -> Value;
    pub fn rb_ary_sum(argc: c_int, argv: *Value, ary: Value) -> Value;
```

**Comment by nomoon** [comment]
https://github.com/ruby/ruby/blob/trunk/array.c#L6077 might also be helpful for looking over the public methods defined on Array to see within Ruby to see if any missing ones stick out.

**Comment by d-unsed** [comment]
@nomoon, could you create a PR for these functions?

**Comment by nomoon** [comment]
@d-unseductable I'd like to do something a bit more comprehensive (i.e. the rest of the classes, not just Array). However, owing to the not-so-great documentation of the Ruby API, I'm not completely sure what's a good or bad idea to submit.

I guess it's C and any of these functions are theoretically accessible? Or should I limit things to what's described by ruby.h and the header files it includes?

**Comment by nomoon** [comment]
@d-unseductable Quite a few of the functions in array.c are not in the header files. Most notably perhaps is `count` and `length`, which I think you've implemented on your side as `rb_ary_len`. So that suggests we should not provide those builtins to ruby-sys, since they wouldn't be available to a C extension compiling against the existing headers (I'm looking at the 2.3 header files).

**Comment by d-unsed** [comment]
@nomoon you're right, only those functions which are present in `ruby.h` and `intern.h` are accessible from `ruby-sys`

**Pull request merged**